### PR TITLE
[net6][msbuild] Set SelfContained for AppExtensions sooner to fix publishing

### PR DESCRIPTION
**Pull request by chamons**
This PR resolves a crash when running the linker on publishing iOS extensions.

The crash would occur [here](https://github.com/xamarin/xamarin-macios/blob/143a2216a60873ea530cd304f8a41f14a2f0b8a6/tools/common/StaticRegistrar.cs#L996) in failing to resolve corelib.

The reason this would fail was System.Private.CoreLib.dll was not in input_assemblies.

This was because we were passes the set of reference assemblies not the expected 'real' ones, and those do not include CoreLib.

After a bunch of digging, this was because _ComputeManagedRuntimePackAssembliesIfSelfContained target was not being set as a condition of  _ComputeAssembliesToPostprocessOnPublish.

_ComputeManagedRuntimePackAssembliesIfSelfContained happened to be the place these were added, and wasn't being set since it has a condition of $(SelfContained) == 'true'

Now confusingly SelfContained WAS being set to true, but only in the targets file, which was too late, as it was checked in a 'global' property group outside of a target.

This means we'd fail to set SelfContained until after the condition, and not run.

This was verified by setting /p:SelfContained=true to true.

I also looked at removing the condition above, since https://github.com/dotnet/runtime/issues/54406 is fixed, however this caused project that didn't set RuntimeIdentifier to fail.

### CODE REVIEW

**Review by mandel-macaque — APPROVED** [review]

**Review by jonathanpeppers — COMMENTED** [review]
I'd maybe suggest adding an app extension that is built in `Release` mode on CI? Would that make sure it's fixed going forwrd?

File: dotnet/targets/Xamarin.Shared.Sdk.props [hunk 1]
```diff
@@ -52,6 +52,11 @@
 		<UseAppHost>false</UseAppHost>
 	</PropertyGroup>
 
+	<!-- App extensions are self-contained, even though their OutputType=Library. This must be done here at not targets as it is checked before targets are invoked. -->
```
Comment (rolfbjarne) [inline dotnet/targets/Xamarin.Shared.Sdk.props:55]:
```suggestion
	<!-- App extensions are self-contained, even though their OutputType=Library. This must be done here and not targets as it is checked before targets are invoked. -->
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]
We'll need a test case for this as well:

1. Add two test project to tests/dotnet, one container project and one extension project. This should be done for all our platforms, similar to how it's done in tests/dotnet/MySimpleApp. So something like:

        tests/dotnet/ExtensionProject/
            macOS/ <-- files specific to the macOS extension
            iOS/
            tvOS/
            MacCatalyst/
            <shared files>

    each platform will probably have to be a different type of extension, since I don't think there's an extension type that's available everywhere.

        tests/dotnet/ExtensionConsumer/
            macOS/ <-- files specific to the macOS project
            iOS/
            tvOS/
            MacCatalyst/
            <shared files>

2. Add a unit test to tests/dotnet/UnitTests/ProjectTest.cs

**Review by spouliot — APPROVED** [review]

**Review by emaf — APPROVED** [review]

**Comment by chamons** [comment]
Tests added, but I had to do them without running. I hit the workload issue and was told a full git clean was needed for me to not get the dreaded:

```
 Possible reasons for this include:
   * You intended to execute a .NET program:
       The application 'build' does not exist.
   * You intended to execute a .NET SDK command:
       A compatible installed .NET SDK for global.json version [6.0.100-rc.2.21456.8] from [/Users/donblas/Programming/macios/master/xamarin-macios/tests/dotnet/ExtensionConsumer/macOS/global.json] was not found.
       Install the [6.0.100-rc.2.21456.8] .NET SDK or update [/Users/donblas/Programming/macios/master/xamarin-macios/tests/dotnet/ExtensionConsumer/macOS/global.json] with an installed .NET SDK:
         6.0.100-rc.2.21423.21 [/Users/donblas/Programming/macios/master/xamarin-macios/builds/downloads/dotnet-sdk-6.0.100-rc.2.21423.21-osx-x64/sdk]
```

**Comment by chamons** [comment]
Tests are failing, but I can run them now then I did a full build...

File: tests/dotnet/ExtensionProject/tvOS/MyTVServicesExtension.csproj [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net6.0-tvos</TargetFramework>
+    <RuntimeIdentifier>tvossimulator-x64</RuntimeIdentifier>
```
Comment (rolfbjarne) [inline tests/dotnet/ExtensionProject/tvOS/MyTVServicesExtension.csproj:5]: Same:

```suggestion
```

File: tests/dotnet/UnitTests/ProjectTest.cs [hunk 3]
```diff
@@ -747,5 +747,29 @@ public void KillEverything ()
 			// https://github.com/xamarin/xamarin-macios/issues/10012
 			ExecutionHelper.Execute ("xcrun", new [] { "simctl", "list" });
 		}
+
+
+		[TestCase (ApplePlatform.iOS, "MyShareExtension.csproj", "bin/Debug/net6.0-ios/iossimulator-x64/MySimpleApp.app/PlugIns/MyShareExtension.appex")]
+		[TestCase (ApplePlatform.TVOS, "MyTVServicesExtension.csproj", "bin/Debug/net6.0-tvos/tvossimulator-x64/MySimpleApp.app/PlugIns/MyTVServicesExtension.appex")]		
+		[TestCase (ApplePlatform.MacOSX, "ShareExtensionTest.csproj", "bin/Debug/net6.0-macos/osx-x64/MySimpleApp.app/Contents/PlugIns/ShareExtensionTest.appex")]
```
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/ProjectTest.cs:754]: As mentioned before, if you name all the extension projects the same way, you can the platform-specific special casing here.

File: tests/dotnet/UnitTests/ProjectTest.cs [hunk 4]
```diff
@@ -747,5 +747,29 @@ public void KillEverything ()
...
+		{
+			Configuration.IgnoreIfIgnoredPlatform (platform);
+			var consumingProjectDir = GetProjectPath ("ExtensionConsumer", platform: platform);
+			var extensionProjectDir = GetProjectPath ("ExtensionProject", platform: platform);
+
+			Clean (extensionProjectDir);
+			Clean (consumingProjectDir);
+
+			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName(extensionProjectDir));
+			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName(consumingProjectDir));
```
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/ProjectTest.cs:766]:
```suggestion
			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName (extensionProjectDir));
			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName (consumingProjectDir));
```

File: tests/dotnet/UnitTests/ProjectTest.cs [hunk 5]
```diff
@@ -747,5 +747,29 @@ public void KillEverything ()
...
+
+			DotNet.AssertBuild (extensionProjectDir, verbosity);
+			DotNet.AssertBuild (consumingProjectDir, verbosity);
+			
+			var extensionPath = Path.Combine (Path.GetDirectoryName(consumingProjectDir), appPath);
```
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/ProjectTest.cs:771]:
```suggestion
			var extensionPath = Path.Combine (Path.GetDirectoryName (consumingProjectDir), appPath);
```

File: tests/dotnet/UnitTests/ProjectTest.cs [hunk 6]
```diff
@@ -747,5 +747,29 @@ public void KillEverything ()
...
+			Assert.That (Directory.Exists(extensionPath), $"App extension directory does not exist: {extensionPath}");
```
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/ProjectTest.cs:772]:
```suggestion
			Assert.That (Directory.Exists (extensionPath), $"App extension directory does not exist: {extensionPath}");
```

File: tests/dotnet/UnitTests/ProjectTest.cs [hunk 7]
```diff
@@ -747,5 +747,29 @@ public void KillEverything ()
...
+			var consumingProjectDir = GetProjectPath ("ExtensionConsumer", platform: platform);
+			var extensionProjectDir = GetProjectPath ("ExtensionProject", platform: platform);
+
+			Clean (extensionProjectDir);
+			Clean (consumingProjectDir);
+
+			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName(extensionProjectDir));
+			Configuration.CopyDotNetSupportingFiles (Path.GetDirectoryName(consumingProjectDir));
+
+			DotNet.AssertBuild (extensionProjectDir, verbosity);
```
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/ProjectTest.cs:768]: You shouldn't have to build the extension project manually first, just building the consuming project should be enough.

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: tests/dotnet/ExtensionProject/iOS/ShareViewController.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+			// Release any cached data, images, etc that aren't in use.
+		}
+
+		public override void ViewDidLoad ()
+		{
+			base.ViewDidLoad ();
+
+			// Do any additional setup after loading the view.
+		}
+
```
[the remark this answers is not in the document]
Comment (chamons) [reply tests/dotnet/ExtensionProject/iOS/ShareViewController.cs:30]: I literally copy pasted this from an exisitng test, but you are right.

**Comment by chamons** [comment]
@rolfbjarne - I believe I have all of your requested test changes in.

**Review by dalexsoto — APPROVED** [review]

File: tests/dotnet/ExtensionConsumer/shared.csproj [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+	<Import Project="../../common/shared-dotnet.csproj" />
+
+	<ItemGroup>
+		<Compile Include="../*.cs" />
+	</ItemGroup>
+
+	<ItemGroup>
+	    <ProjectReference Include="..\..\ExtensionProject\$(_PlatformName)\ExtensionProject.csproj">
+    	  <IsAppExtension>true</IsAppExtension>
+    	</ProjectReference>
```
Comment (rolfbjarne) [inline tests/dotnet/ExtensionConsumer/shared.csproj:20]: Minor: indentation is off here.

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**